Source/ExecutionEngine/ExecutionEngine.cs

```diff
@@ -10,175 +10,17 @@
 using BoogiePL = Microsoft.Boogie;
 using System.Runtime.Caching;
 using System.Diagnostics;
-using System.Net.Mime;
+using System.Runtime.CompilerServices;
+using VCGeneration;
 
 namespace Microsoft.Boogie
 {
-  #region Output printing
-
-  public interface OutputPrinter : IConditionGenerationPrinter
-  {
-    ExecutionEngineOptions Options { get; set; }
-    void ErrorWriteLine(TextWriter tw, string s);
-    void ErrorWriteLine(TextWriter tw, string format, params object[] args);
-    void AdvisoryWriteLine(TextWriter output, string format, params object[] args);
-    void Inform(string s, TextWriter tw);
-    void WriteTrailer(TextWriter textWriter, PipelineStatistics stats);
-    void WriteErrorInformation(ErrorInformation errorInfo, TextWriter tw, bool skipExecutionTrace = true);
-    void ReportBplError(IToken tok, string message, bool error, TextWriter tw, string category = null);
-
-    void ReportImplementationsBeforeVerification(Implementation[] implementations);
-    void ReportStartVerifyImplementation(Implementation implementation);
-    void ReportEndVerifyImplementation(Implementation implementation, VerificationResult result);
-  }
-
-  #endregion
-
-  public enum PipelineOutcome
-  {
-    Done,
-    ResolutionError,
-    TypeCheckingError,
-    ResolvedAndTypeChecked,
-    FatalError,
-    Cancelled,
-    VerificationCompleted
-  }
-
-
-  public class PipelineStatistics
-  {
-    public int ErrorCount;
-    public int VerifiedCount;
-    public int InconclusiveCount;
-    public int TimeoutCount;
-    public int OutOfResourceCount;
-    public int OutOfMemoryCount;
-    public int SolverExceptionCount;
-    public long[] CachingActionCounts;
-    public int CachedErrorCount;
-    public int CachedVerifiedCount;
-    public int CachedInconclusiveCount;
-    public int CachedTimeoutCount;
-    public int CachedOutOfResourceCount;
-    public int CachedOutOfMemoryCount;
-    public int CachedSolverExceptionCount;
-  }
-
-
-  #region Error reporting
-
-  public delegate void ErrorReporterDelegate(ErrorInformation errInfo);
-
-
-  public enum ErrorKind
-  {
-    Assertion,
-    Precondition,
-    Postcondition,
-    InvariantEntry,
-    InvariantMaintainance
-  }
-
-
-  public class ErrorInformationFactory
-  {
-    public virtual ErrorInformation CreateErrorInformation(IToken tok, string msg, string requestId = null,
-      string originalRequestId = null, string category = null)
-    {
-      Contract.Requires(1 <= tok.line && 1 <= tok.col);
-      Contract.Requires(msg != null);
-
-      return ErrorInformation.CreateErrorInformation(tok, msg, requestId, originalRequestId, category);
+  public record ProcessedProgram(Program Program, Action<VCGen, Implementation, VerificationResult> PostProcessResult) {
```

Comment:
I like the last one. The first one I'm not a fan of because I don't like widening a type when the original type is sufficient. The second one is I think slightly less future proof than the current option, since you can imagine a function that takes a `ProcessedProgram` and further processes it to return another `ProcessedProgram`, but that would be difficult when `ProcessedProgram` is already a specific subclass. The third one is more specific than the current solution which I like, but the name exposes slightly more details than I'd like to.